common/reconciliation/invariants/historyExists.go

```diff
@@ -49,17 +49,25 @@ func NewHistoryExists(
 	}
 }
 
-func (h *historyExists) Check(execution common.Execution) common.CheckResult {
+func (h *historyExists) Check(execution interface{}) common.CheckResult {
```

Comment:
This method should take in common.ConcreteExectuion. Invariants for concrete execution cannot be applied on current execution (and the reverse is also true) so at this level in code I don't think it makes sense to still be dealing with interfaces (or even dealing with union types at this level if we go with that path).

Reply:
here and below: could you please elaborate your proposal with a small design example. Otherwise, let's not overcomplicate the current one.

Reply:
My only point here is that at this level in code you should not be dealing with an interface. Invariant checks for concrete executions do not have to do with current execution checks so at the level of invariants we should not be taking interfaces.